### DISCUSSION TITLE
would velero provider restful api for backup&restore actions.

### DISCUSSION THREAD

**Discussion by wenterjoy** [vmware-tanzu/velero#5754] [Community support Q&A] [answered]
I want involve velero as backup and restore tools into our PAAS platform. so I would develop some restful api for velero. so it can be invoked by other components. Does this in velero plan? or for some other obvious consideration，this is not velero‘s direction？

**Comment by wenterjoy** [comment]
Hi Team,

Can someone give some reply.

Thanks,

**Reply by draghuram** [reply]
This question is best answered by a core team member but I don't remember seeing any discussions in this regard. Do note that Velero is managed by creating Kubernetes resources and as such, Kubernetes API can be used to control Velero. But that is not nearly same as generic REST API.

**Reply by wenterjoy** [reply]
thanks a lot for reply. @draghuram 
Kubernetes API can be used in cluster internally.
If external users or ends wants to invoke velero's function in programming mode，calling console commands may not be the better way.
This is why i want to implement the REST API.

**Comment by sseago** [comment, marked as answer]
To support this, you'd probably need to build some sort of server or controller that created velero backup or restore CRs in response to REST actions. The way the Velero server works is that the backup and restore controllers initiate backup or restore actions in response toe the creation of Velero kubernetes custom resources.

**Reply by wenterjoy** [reply]
thanks for the answer. @sseago 
Yes. It would expose a http server to accept request, but the server should call the velero client factory to invoke the backup or restore actions as velero internally does. By using the velero client factory should save a lot of repetitive efforts. In indeed, I have enabled a server that can get, create backups and logs in this way( with most parameters with default values.)
From your answer , should I think it not the direction that velero will definitely not enter? I would continue on it and open it to velero community.

**Reply by sseago** [reply]
So we'd still need the velero CR created, since everything else depends on it. This new HTTP server would just need to create the velero CR, and the existing backup controller would start the backup just as if the user had created the CR manually or if the velero CLI had created it.

**Reply by wenterjoy** [reply]
Yes. Your description is more accurate. Thanks a lot.
 I would continue.
